docs/source/cuda/examples.rst

```diff
@@ -3,6 +3,479 @@
+scope for this documentation, but it will suffice to say that it describes how heat
+propagates through an object over time. It works by discretizing the problem in two ways:
+
+1. The object is separated into small "pieces" that each have an individual temperature
+2. Time is separated into small "intervals" and the universe advances through them one by one
+
+Then, the following assumption is applied: The temperature of a piece after some interval 
+has passed is some weighted average of the temperature of the pieces that are directly 
+touching it. Intuitively, this means if all the small pieces of the object are very hot 
+and a single piece in the middle is very cold, as time goes on, the hot pieces will cause 
+the cold one to heat up and the cold piece will cause the surrounding hot pieces to cool 
+slightly. Simply put, the heat spreads out around the object.
+
+We can simulate this situation using a Kernel generated from Numba. Let's start simple by assuming
+we have a one dimensional object which we'll represent with an array of values. The position 
+of the element in the array is like the position of a small piece of the object, and the value 
+of the element is like the temperature. 
+
+.. code-block:: python
+
+   from numba import cuda
+   import numpy as np
+
+
+Some initial setup here. Let's make one point in the center of the object very hot.
+
+.. code-block:: python
+
+   size = 100
+   data = np.zeros(size)
+
+   # make one piece in the very middle very hot
+   data[int(size/2)] = 10000
+   data_gpu = cuda.to_device(data)
+
+
+In our kernel each thread will be responsible for managing the temperature update for a single element
+in a loop over the desired number of timesteps. The kernel is below:
+
+.. code-block:: python
+
+   @cuda.jit
+   def solve_heat_equation(data, size, timesteps, k):
+      i = cuda.grid(1)
+
+      for step in range(timesteps):
+         # get the current temperature associated with this segment
+         curr_temp = data[i]
+         
+         # apply formula from finite difference equation
+         if i == 0:
+               # Left wall is held at T = 0
+               next_temp = curr_temp + k * (data[i + 1] - (2 * curr_temp))
+         elif i == size - 1:
+               # Right wall is held at T = 0
+               next_temp = curr_temp + k * (data[i - 1] - (2 * curr_temp))
+         else:
+               next_temp = curr_temp + k * (data[i - 1] - (2 * curr_temp) + data[i + 1])
+         data[i] = next_temp
+         
+         # wait for every thread to write before moving on
+         cuda.syncthreads()
+
+Calling the kernel:
+
+.. code-block:: python
+
+   solve_heat_equation.forall(len(data))(data_gpu, len(data_gpu), 10000, 0.25)
+
+Since the kernel reused the original data vector to work on the problem, copying it back
+to the host yields the results:
+
+.. code-block:: python
+
+   print(data_gpu.copy_to_host())
```

Comment:
It might be nice to include before and after plots generated with something like:

```python
import matplotlib.pyplot as plt

plt.plot(np.arange(len(data)), data_gpu.copy_to_host())
plt.title('Heat in a bar')
plt.xlabel('Position')
plt.ylabel('Temperature')
plt.show()
```

to give something like:

![temp](https://user-images.githubusercontent.com/535640/161045589-8c9c2081-7add-4541-a195-ce8ec4e34f8e.png)

Reply:
I really wanted to include some plots here but didn't see any precedent for that elsewhere in the docs. Happy to add them in though.